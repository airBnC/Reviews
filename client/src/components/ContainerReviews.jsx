import React from 'react';
import PropTypes from 'prop-types';

import ReviewEntry from './ReviewEntry';

import { BasicWrapperFlexColumnDiv } from './styles/MasterStyles.styles';

const ContainerReviews = props => (
  <BasicWrapperFlexColumnDiv>
    {props.reviews.map(review => <ReviewEntry key={review.key} review={review} />)}
  </BasicWrapperFlexColumnDiv>
);

ContainerReviews.propTypes = {
  reviews: PropTypes.arrayOf(PropTypes.shape({
    review_date: PropTypes.string,
    review_text: PropTypes.string,
    first_name: PropTypes.string,
    last_name: PropTypes.string,
    key: PropTypes.number,
  })),
};

ContainerReviews.defaultProps = {
  reviews: [{
    review_date: '2016-11-04',
    review_text: 'Laborum eius id rerum et in ratione',
    first_name: 'Toby',
    last_name: 'Keith',
    key: 0,
  }],
};

export default ContainerReviews;
