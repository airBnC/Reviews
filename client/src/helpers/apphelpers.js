export const filterReviews = (searchString, processedReviewsArray) => {
  const compareString = searchString.toLowerCase();
  return processedReviewsArray.filter((reviewObj) => {
    return reviewObj.review_text.toLowerCase().includes(compareString);
  });
};

// -----------------
// -----------------
// -----------------

// Note:
// I used the below formulas/objects when I was handling data reformatting on the client side
// But I then decided to do formatting on the server side.
// However, I'm leaving this here just in case I revisit that decision in the future.

/*
const months = {
  1: 'January',
  2: 'February',
  3: 'March',
  4: 'April',
  5: 'May',
  6: 'June',
  7: 'July',
  8: 'August',
  9: 'September',
  10: 'October',
  11: 'November',
  12: 'December',
};

const trimDateToYYYYMMDD = date => date.slice(0, 11);

const convertDateToMonthYYYY = (trimmedDate) => {
  const month = +trimmedDate.slice(5, 7);
  const year = trimmedDate.slice(0, 4);
  return `${months[month]} ${year}`;
};

export const processReviewsArray = (reviewsArray) => {
  const reviewsArrayCopy = reviewsArray.slice();
  let counter = 0;
  reviewsArrayCopy.forEach((review) => {
    review.review_date = convertDateToMonthYYYY(trimDateToYYYYMMDD(review.review_date));
    review.key = counter;
    counter += 1;
  });
  return reviewsArrayCopy;
};
*/
