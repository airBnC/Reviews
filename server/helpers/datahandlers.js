module.exports.calcReviewsAverageStars = (averageStarsArray) => {
  let accuracyTot = 0;
  let communicationTot = 0;
  let cleanlinessTot = 0;
  let locationTot = 0;
  let checkinTot = 0;
  let valueTot = 0;
  const numReviews = averageStarsArray.length;

  averageStarsArray.forEach((reviewsStarsObj) => {
    accuracyTot += reviewsStarsObj.rank_accuracy;
    communicationTot += reviewsStarsObj.rank_communication;
    cleanlinessTot += reviewsStarsObj.rank_cleanliness;
    locationTot += reviewsStarsObj.rank_location;
    checkinTot += reviewsStarsObj.rank_checkin;
    valueTot += reviewsStarsObj.rank_value;
  });

  const accuracyAvg = +(accuracyTot / numReviews).toFixed(2);
  const communicationAvg = +(communicationTot / numReviews).toFixed(2);
  const cleanlinessAvg = +(cleanlinessTot / numReviews).toFixed(2);
  const locationAvg = +(locationTot / numReviews).toFixed(2);
  const checkinAvg = +(checkinTot / numReviews).toFixed(2);
  const valueAvg = +(valueTot / numReviews).toFixed(2);
  const overallAvg = +((accuracyAvg + communicationAvg + cleanlinessAvg + locationAvg + checkinAvg + valueAvg) / 6)
    .toFixed(2);

  return {
    overallAvg,
    accuracyAvg,
    communicationAvg,
    cleanlinessAvg,
    locationAvg,
    checkinAvg,
    valueAvg,
    numReviews,
  };
};

// ------------
// ------------
// ------------

module.exports.checkForValidRecord = (queryId, totalRecords) => {
  return (queryId > 0 && queryId <= totalRecords) ? true : false;
};

// ------------
// ------------
// ------------

const months = {
  1: 'January',
  2: 'February',
  3: 'March',
  4: 'April',
  5: 'May',
  6: 'June',
  7: 'July',
  8: 'August',
  9: 'September',
  10: 'October',
  11: 'November',
  12: 'December',
};

const convertDateToMonthYYYY = (trimmedDate) => {
  const month = +trimmedDate.slice(0, 2);
  const year = trimmedDate.slice(6, 10);
  return `${months[month]} ${year}`;
};

module.exports.processReviewsArray = (reviewsArray) => {
  const reviewsArrayCopy = reviewsArray.slice();
  const options = { year: 'numeric', month: '2-digit', day: '2-digit' };
  reviewsArrayCopy.forEach((review) => {
    review.review_date = convertDateToMonthYYYY(review.review_date.toLocaleDateString('en-US', options));
  });
  return reviewsArrayCopy;
};
