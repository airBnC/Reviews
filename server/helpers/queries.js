const db = require('../../database/index.js');
const mysql = require('mysql');
const Promise = require('bluebird');

module.exports.listingAverageStars = {
  get: (listingId) => {
    return new Promise((resolve, reject) => {
      // Using prepared statements to protect against SQL Injection Attacks
      // See here: https://dev.mysql.com/doc/refman/5.7/en/sql-syntax-prepared-statements.html
      // And here: https://github.com/mysqljs/mysql#preparing-queries
      const sql = `SELECT AVG(rank_accuracy) as rank_accuracy, AVG(rank_communication) as rank_communication, 
                 AVG(rank_cleanliness) as rank_cleanliness, AVG(rank_location) as rank_location, 
                 AVG(rank_checkin) as rank_checkin, AVG(rank_value) as rank_value 
                 FROM tbl_reviews WHERE listing_id = $1`;
      db.pgConnection.query(sql, [listingId], (err, result) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
  },
};

module.exports.listingReviews = {
  get: (listingId) => {
    return new Promise((resolve, reject) => {
      const sql = `SELECT r.review_date, r.review_text, r.first_name 
                   FROM tbl_reviews as r 
                  WHERE listing_id = $1`;
      db.pgConnection.query(sql, [listingId], (err, result) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
  },
  post: (review) => {
    return new Promise((resolve, reject) => {
      let sql = 'INSERT INTO tbl_reviews (listing_id, user_id, review_date, review_text, rank_accuracy, rank_communication, rank_cleanliness, rank_location, rank_checkin, rank_value) VALES (??, ??, ??, ??, ??, ??, ??, ??, ??, ??)';
      const listingId = review.listingId;
      const userId = review.userId;
      const reviewDate = review.reviewDate || new Date();
      const reviewText = review.reviewText || '';
      const rankAccuracy = review.rankAccuracy || 0;
      const rankCommunication = review.rankCommunication || 0;
      const rankCleanliness = review.rankCleanliness || 0;
      const rankLocation = review.rankLocation || 0;
      const rankCheckin = review.rankCheckin || 0;
      const rankValue = review.rankValue || 0;
      const inserts = [listingId, userId, reviewDate, reviewText, rankAccuracy, rankCommunication, rankCleanliness, rankLocation, rankCheckin, rankValue];
      sql = mysql.format(sql, inserts);

      db.dbConnection.query(sql, (err, result) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
  },
  put: (review) => {
    return new Promise((resolve, reject) => {
      const sql = 'SELECT "UPDATE STATEMENT RAN";'; // Just putting in a dummy query in place of an update

      db.dbConnection.query(sql, (err, result) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
  },
  delete: (review) => {
    return new Promise((resolve, reject) => {
      let sql = 'DELETE FROM tbl_review WHERE id = ??;';
      const reviewId = review.reviewId || null;
      const inserts = [reviewId];
      sql = mysql.format(sql, inserts);

      db.dbConnection.query(sql, (err, result) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
  },
};


module.exports.listings = {
  getTotal: () => {
    // Hard-coded to 100 because I don't have actual listings in my database, just references to listings.

    return 100;

    // However, see below for a theoretical representation of how the query might look
    // (assuming I had listing records in a table called 'listing').
    /*
    return new Promise((resolve, reject) => {
      let numReviewsQuery = 'SELECT COUNT(*) FROM ??';
      const inserts = ['listing'];
      numReviewsQuery = mysql.format(starsQuery, inserts);

      db.dbConnection.query(numReviewsQuery, (err, result) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
    */
  },
};
