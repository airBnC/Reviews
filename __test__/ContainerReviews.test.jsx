import React from 'react';
import { configure, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import ContainerReviews from '../client/src/components/ContainerReviews';

import { processReviewsArray } from '../client/src/helpers/apphelpers';

configure({ adapter: new Adapter() });

const processedDummyReviews = [
  {
    review_date: 'February 2017',
    review_text: 'Debitis sequi id nam optio iste molestias dolores dolorem ex.',
    first_name: 'Peggie',
    last_name: 'Oberbrunner',
    key: 0,
  },
  {
    review_date: 'October 2016',
    review_text: 'Dolore doloribus eius numquam sunt sapiente quos officia quod nulla.',
    first_name: 'Jasmin',
    last_name: 'Schmidt',
    key: 1,
  },
];

describe('ContainerReviews', () => {
  it('should render correctly', () => {
    const output = shallow(<ContainerReviews reviews={processedDummyReviews} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
