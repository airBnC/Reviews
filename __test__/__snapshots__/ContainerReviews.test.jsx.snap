// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContainerReviews should render correctly 1`] = `
<styled.div>
  <ReviewEntry
    key="0"
    review={
      Object {
        "first_name": "Peggie",
        "key": 0,
        "last_name": "Oberbrunner",
        "review_date": "February 2017",
        "review_text": "Debitis sequi id nam optio iste molestias dolores dolorem ex.",
      }
    }
  />
  <ReviewEntry
    key="1"
    review={
      Object {
        "first_name": "Jasmin",
        "key": 1,
        "last_name": "Schmidt",
        "review_date": "October 2016",
        "review_text": "Dolore doloribus eius numquam sunt sapiente quos officia quod nulla.",
      }
    }
  />
</styled.div>
`;
